File: Pharmaceutical_Company_Example/03_Behavioral/R/Template_Method.R
# Define Abstract Class with Template Method
RegulatoryProcess <- setRefClass(
  "RegulatoryProcess",
  methods = list(
    # Template Method
    execute_process = function(drug_name) {
      message("Starting the regulatory process for drug:", drug_name)
      perform_evaluation(drug_name)
      perform_approval(drug_name)
      finalize_process(drug_name)
      message("Regulatory process completed for drug:", drug_name)
    },
    
    # Abstract Methods
    perform_evaluation = function(drug_name) {
      stop("perform_evaluation method should be overridden by concrete subclasses!")
    },
    
    perform_approval = function(drug_name) {
      stop("perform_approval method should be overridden by concrete subclasses!")
    },
    
    finalize_process = function(drug_name) {
      message("Finalizing process for drug:", drug_name)
    }
  )
)

# Concrete Class: Standard Approval Process
StandardApprovalProcess <- setRefClass(
  "StandardApprovalProcess",
  contains = "RegulatoryProcess",
  methods = list(
    perform_evaluation = function(drug_name) {
      message("Evaluating drug:", drug_name, "using Standard Evaluation.")
    },
    
    perform_approval = function(drug_name) {
      message("Approving drug:", drug_name, "using Standard Approval Process.")
    }
  )
)

# Concrete Class: Accelerated Approval Process
AcceleratedApprovalProcess <- setRefClass(
  "AcceleratedApprovalProcess",
  contains = "RegulatoryProcess",
  methods = list(
    perform_evaluation = function(drug_name) {
      message("Evaluating drug:", drug_name, "using Accelerated Evaluation.")
    },
    
    perform_approval = function(drug_name) {
      message("Approving drug:", drug_name, "using Accelerated Approval Process.")
    }
  )
)

# Create instances of the concrete classes
standard_process <- StandardApprovalProcess$new()
accelerated_process <- AcceleratedApprovalProcess$new()

# Execute processes
standard_process$execute_process("Test Small Molecule 01")
accelerated_process$execute_process("Test Small Molecule 02")

File: Pharmaceutical_Company_Example/03_Behavioral/PHP/Template_Method.php
<?php

// Abstract Class with Template Method
abstract class RegulatoryProcess
{
    public function executeProcess(string $drugName): void
    {
        echo "Starting the regulatory process for drug: $drugName\n";
        $this->performEvaluation($drugName);
        $this->performApproval($drugName);
        $this->finalizeProcess($drugName);
        echo "Regulatory process completed for drug: $drugName\n";
    }

    // Abstract Methods
    abstract protected function performEvaluation(string $drugName): void;
    abstract protected function performApproval(string $drugName): void;

    // Final Method
    protected function finalizeProcess(string $drugName): void
    {
        echo "Finalizing process for drug: $drugName\n";
    }
}

// Concrete Class: Standard Approval Process
class StandardApprovalProcess extends RegulatoryProcess
{
    protected function performEvaluation(string $drugName): void
    {
        echo "Evaluating drug: $drugName using Standard Evaluation.\n";
    }

    protected function performApproval(string $drugName): void
    {
        echo "Approving drug: $drugName using Standard Approval Process.\n";
    }
}

// Concrete Class: Accelerated Approval Process
class AcceleratedApprovalProcess extends RegulatoryProcess
{
    protected function performEvaluation(string $drugName): void
    {
        echo "Evaluating drug: $drugName using Accelerated Evaluation.\n";
    }

    protected function performApproval(string $drugName): void
    {
        echo "Approving drug: $drugName using Accelerated Approval Process.\n";
    }
}

// Create instances of the concrete classes
$standardProcess = new StandardApprovalProcess();
$acceleratedProcess = new AcceleratedApprovalProcess();

// Execute processes
$standardProcess->executeProcess("Test Small Molecule 01");
$acceleratedProcess->executeProcess("Test Small Molecule 02");

?>
